clippy_lints/src/types.rs

```diff
@@ -1100,6 +1100,31 @@ declare_clippy_lint! {
     "casting a function pointer to a numeric type other than usize"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for casts of function directly to a usize
+    ///
+    /// **Why is this bad?**
+    /// Casting a function directly to a usize is likely a mistake. Most likely, the intended behavior
+    /// is calling the function. If casting to a usize is intended, it would be more clearly expressed
+    /// by casting to a function pointer, then to a usize.
+    ///
+    /// **Example**
+    ///
+    /// ```rust
+    /// // Bad
+    /// fn fun() -> usize { 1 }
+    /// let a: usize = fun as usize;
+    ///
+    /// // Good
+    /// fn fun2() -> usize { 1 }
+    /// let a: usize = fun2 as fn() -> usize as usize;
+    /// let b: usize = fun2();
+    /// ```
+    pub FN_TO_NUMERIC_CAST_USIZE,
+    style,
```

Comment:
This should be `pedantic` or even `restriction` IMO, since this cast is usually fine.